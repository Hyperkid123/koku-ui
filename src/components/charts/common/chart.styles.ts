import {
  global_Color_dark_100,
  global_Color_light_100,
  global_primary_color_100,
  global_primary_color_200,
  global_success_color_100,
  global_success_color_200,
} from '@patternfly/react-tokens';
import { css } from 'emotion';
import React from 'react';

export const chartStyles = {
  padding: 8,
  group: {
    data: { strokeWidth: 2, fillOpacity: 0.4 },
  },
  tooltipText: {
    fontSize: '14px',
    fill: global_Color_light_100.value,
  } as React.CSSProperties,
  tooltipFlyout: { fill: global_Color_dark_100.value } as React.CSSProperties,
  previousMonth: {
    data: {
      fill: global_success_color_200.value,
      stroke: global_success_color_100.value,
    },
  },
  currentMonth: {
    data: {
      fill: global_primary_color_100.value,
      stroke: global_primary_color_200.value,
    },
  },
};

export const chartOverride = css`
  :not(foo) svg {
    overflow: visible;
  }
`;
