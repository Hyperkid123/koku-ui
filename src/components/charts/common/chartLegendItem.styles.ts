import {
  global_BorderWidth_sm,
  global_Color_200,
  global_FontSize_xs,
  global_spacer_md,
  global_spacer_sm,
} from '@patternfly/react-tokens';
import React from 'react';
import { chartStyles } from './chart.styles';

export const styles = {
  legendItem: {
    display: 'flex',
    alignItems: 'center',
    fontSize: global_FontSize_xs.value,
    color: global_Color_200.var,
    marginRight: global_spacer_md.value,
  },
  color: {
    height: 10,
    width: 10,
    marginRight: global_spacer_sm.value,
    borderWidth: global_BorderWidth_sm.value,
    borderStyle: 'solid',
  },
  currentColor: {
    backgroundColor: chartStyles.currentMonth.data.fill,
    borderColor: chartStyles.currentMonth.data.stroke,
  },
  previousColor: {
    backgroundColor: chartStyles.previousMonth.data.fill,
    borderColor: chartStyles.previousMonth.data.stroke,
  },
} as { [className: string]: React.CSSProperties };
