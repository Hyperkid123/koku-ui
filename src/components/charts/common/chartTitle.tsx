import { css } from '@patternfly/react-styles';
import { global_FontSize_xs, global_spacer_sm } from '@patternfly/react-tokens';
import React from 'react';

const styles = {
  chartTitle: {
    fontSize: global_FontSize_xs.value,
    marginBottom: global_spacer_sm.value,
  },
} as { [className: string]: React.CSSProperties };

interface ChartTitleProps {
  children: React.ReactNode;
}

const ChartTitle: React.SFC<ChartTitleProps> = ({ children }) => (
  <div className={css(styles.chartTitle)}>{children}</div>
);

export { ChartTitle, ChartTitleProps };
