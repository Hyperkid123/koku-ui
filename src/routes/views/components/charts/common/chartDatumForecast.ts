import type { Forecast } from 'api/forecasts/forecast';
import { getDate } from 'date-fns';
import type { ComputedForecastItem } from 'utils/computedForecast/getComputedForecastItems';
import { getComputedForecastItems } from 'utils/computedForecast/getComputedForecastItems';
import { SortDirection } from 'utils/sort';

import type { ChartDatum } from './chartDatum';
import { DatumType, isFloat, isInt, padChartDatums } from './chartDatum';

// The computed forecast cost
// eslint-disable-next-line no-shadow
export const enum ComputedForecastItemType {
  cost = 'cost',
  infrastructure = 'infrastructure',
  supplementary = 'supplementary',
}

export function transformForecast(
  forecast: Forecast,
  datumType,
  forecastItem: string = 'cost',
  forecastItemValue: string = 'total'
): ChartDatum[] {
  if (!forecast) {
    return [];
  }
  const items = {
    forecast,
    sortKey: 'date',
    sortDirection: SortDirection.desc,
  } as any;
  const computedItems = getComputedForecastItems(items);
  let chartDatums;
  if (datumType === DatumType.cumulative) {
    chartDatums = computedItems.reduce<ChartDatum[]>((acc, d) => {
      const prevValue = acc.length ? acc[acc.length - 1].y : 0;
      return [...acc, createForecastDatum(prevValue + d[forecastItem][forecastItemValue].value, d)];
    }, []);
  } else {
    chartDatums = computedItems.map(i => createForecastDatum(i[forecastItem][forecastItemValue].value, i));
  }
  return padChartDatums(chartDatums, datumType);
}

export function transformForecastCone(
  forecast: Forecast,
  datumType,
  forecastItem: string = 'cost',
  forecastItemValue: string = 'total'
): ChartDatum[] {
  if (!forecast) {
    return [];
  }
  const items = {
    forecast,
    sortKey: 'date',
    sortDirection: SortDirection.desc,
  } as any;
  const computedItems = getComputedForecastItems(items);
  let chartDatums;
  if (datumType === DatumType.cumulative) {
    chartDatums = computedItems.reduce<ChartDatum[]>((acc, d) => {
      const prevMaxValue = acc.length ? acc[acc.length - 1].y : d[forecastItem][forecastItemValue].value;
      const prevMinValue = acc.length ? acc[acc.length - 1].y0 : d[forecastItem][forecastItemValue].value;
      return [
        ...acc,
        createForecastConeDatum(
          prevMaxValue + d[forecastItem].confidence_max.value,
          prevMinValue + d[forecastItem].confidence_min.value,
          d
        ),
      ];
    }, []);
  } else {
    chartDatums = computedItems.map(i =>
      createForecastConeDatum(i[forecastItem].confidence_max.value, i[forecastItem].confidence_min.value, i)
    );
  }
  return padChartDatums(chartDatums, datumType);
}

export function createForecastDatum<T extends ComputedForecastItem>(
  value: number,
  computedItem: T,
  forecastItem: string = 'cost',
  forecastItemValue: string = 'total'
): ChartDatum {
  const xVal = getDate(new Date(computedItem.date + 'T00:00:00'));
  const yVal = isFloat(value) ? parseFloat(value.toFixed(2)) : isInt(value) ? value : 0;
  return {
    x: xVal,
    y: value === null ? null : yVal, // For displaying "no data" labels in chart tooltips
    key: computedItem.date,
    units:
      computedItem[forecastItem] && computedItem[forecastItem][forecastItemValue]
        ? computedItem[forecastItem][forecastItemValue].units
        : undefined,
  };
}

export function createForecastConeDatum<T extends ComputedForecastItem>(
  maxValue: number,
  minValue: number,
  computedItem: T,
  forecastItem: string = 'cost',
  forecastItemValue: string = 'total'
): ChartDatum {
  const xVal = getDate(new Date(computedItem.date + 'T00:00:00'));
  const yVal = isFloat(maxValue) ? parseFloat(maxValue.toFixed(2)) : isInt(maxValue) ? maxValue : 0;
  const y0Val = isFloat(minValue) ? parseFloat(minValue.toFixed(2)) : isInt(minValue) ? minValue : 0;
  return {
    x: xVal,
    y: maxValue === null ? null : yVal, // For displaying "no data" labels in chart tooltips
    y0: minValue === null ? null : y0Val,
    key: computedItem.date,
    units:
      computedItem[forecastItem] && computedItem[forecastItem][forecastItemValue]
        ? computedItem[forecastItem][forecastItemValue].units
        : undefined,
  };
}
