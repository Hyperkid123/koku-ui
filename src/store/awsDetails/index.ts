import * as awsDetailsActions from './awsDetailsActions';
import {
  awsDetailsStateKey,
  AwsDetailsTab,
  AwsDetailsWidget,
} from './awsDetailsCommon';
import { awsDetailsReducer } from './awsDetailsReducer';
import * as awsDetailsSelectors from './awsDetailsSelectors';

export {
  awsDetailsStateKey,
  awsDetailsReducer,
  awsDetailsActions,
  awsDetailsSelectors,
  AwsDetailsTab,
  AwsDetailsWidget,
};
