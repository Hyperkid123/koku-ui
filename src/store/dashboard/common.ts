export interface ValueFormatOptions {
  fractionDigits?: number;
}

export interface DashboardWidget {
  availableTabs?: any[];
  currentTab: any;
  details: {
    costKey?: string /** i18n label key */;
    formatOptions: ValueFormatOptions;
    showUnits?: boolean;
    showUsageLegendLabel?: boolean;
    usageFormatOptions?: ValueFormatOptions;
    usageKey?: string /** i18n label key */;
  };
  filter?: {
    limit?: number;
    service?: string;
  };
  id: number;
  isDetailsLink?: boolean;
  isHorizontal?: boolean;
  isUsageFirst?: boolean;
  reportType: any;
  /** i18n key for the title. passed { startDate, endDate, month, time } */
  titleKey: string;
  tabsFilter?: {
    limit?: number;
    service?: string;
  };
  trend: {
    titleKey: string;
    type: number;
    formatOptions: ValueFormatOptions;
  };
  topItems: {
    formatOptions: {};
  };
}
