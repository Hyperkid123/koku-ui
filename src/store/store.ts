import { notificationsMiddleware } from '@red-hat-insights/insights-frontend-components/components/Notifications';
import axios from 'axios';
import { applyMiddleware, compose, createStore, DeepPartial } from 'redux';
import thunk from 'redux-thunk';
import { rootReducer, RootState } from './rootReducer';
import { sessionActions } from './session';

declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__(options: any): typeof compose;
  }
}

const logger = () => store => next => action => {
  return next(action);
};

const composeEnhancers =
  typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ serialize: true })
    : compose;

export const middlewares = [
  logger(),
  thunk,
  notificationsMiddleware(),
  logger(),
];

export function configureStore(initialState: DeepPartial<RootState>) {
  const enhancer = composeEnhancers(applyMiddleware(...middlewares));
  const store = createStore(rootReducer, initialState, enhancer);

  axios.interceptors.response.use(null, error => {
    if (
      error &&
      error.response &&
      (error.response.status === 403 || error.response.status === 401)
    ) {
      store.dispatch(sessionActions.logout());
    }
    return Promise.reject(error);
  });

  return store;
}
