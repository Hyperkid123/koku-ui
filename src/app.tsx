import { Providers, ProviderType } from 'api/providers';
import { getProvidersQuery } from 'api/providersQuery';
import { AxiosError } from 'axios';
import { I18nProvider } from 'components/i18nProvider';
import React from 'react';
import { hot } from 'react-hot-loader';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router';
import { withRouter } from 'react-router-dom';
import { compose } from 'redux';
import { createMapStateToProps, FetchStatus } from 'store/common';
import {
  awsProvidersQuery,
  ocpProvidersQuery,
  providersActions,
  providersSelectors,
} from 'store/providers';
import { asyncComponent } from 'utils/asyncComponent';
import { Routes } from './routes';
import * as onboardingSelectors from './store/onboarding/selectors';

const ProvidersModal = asyncComponent(() =>
  import(/* webpackChunkName: "providersModal" */ './pages/onboardingModal')
);

export interface AppOwnProps extends RouteComponentProps<void> {}

interface AppStateProps {
  awsProviders: Providers;
  awsProvidersFetchStatus: FetchStatus;
  awsProvidersQueryString: string;
  ocpProviders: Providers;
  ocpProvidersFetchStatus: FetchStatus;
  ocpProvidersQueryString: string;
  onboardingErrors: AxiosError;
  onboardingStatus: FetchStatus;
}

interface AppDispatchProps {
  history: any;
  fetchProviders: typeof providersActions.fetchProviders;
}

interface AppState {
  locale: string;
}

type AppProps = AppOwnProps & AppStateProps & AppDispatchProps;

export class App extends React.Component<AppProps, AppState> {
  public appNav: any;

  public buildNav: any;

  public state: AppState = { locale: 'en' };

  public componentDidMount() {
    const {
      awsProviders,
      awsProvidersFetchStatus,
      awsProvidersQueryString,
      fetchProviders,
      history,
      ocpProviders,
      ocpProvidersFetchStatus,
      ocpProvidersQueryString,
    } = this.props;

    insights.chrome.init();
    insights.chrome.identifyApp('cost-management');
    insights.chrome.navigation(buildNavigation());

    this.appNav = insights.chrome.on('APP_NAVIGATION', event =>
      history.push(`/${event.navId}`)
    );
    this.buildNav = history.listen(() =>
      insights.chrome.navigation(buildNavigation())
    );

    if (!awsProviders && awsProvidersFetchStatus !== FetchStatus.inProgress) {
      fetchProviders(ProviderType.aws, awsProvidersQueryString);
    }
    if (!ocpProviders && ocpProvidersFetchStatus !== FetchStatus.inProgress) {
      fetchProviders(ProviderType.ocp, ocpProvidersQueryString);
    }
  }

  public componentDidUpdate(prevProps: AppProps) {
    const {
      awsProviders,
      awsProvidersFetchStatus,
      awsProvidersQueryString,
      fetchProviders,
      location,
      ocpProviders,
      ocpProvidersFetchStatus,
      ocpProvidersQueryString,
      onboardingErrors,
      onboardingStatus,
    } = this.props;

    if (
      (!awsProviders ||
        (onboardingStatus === FetchStatus.complete && !onboardingErrors)) &&
      awsProvidersFetchStatus !== FetchStatus.inProgress
    ) {
      fetchProviders(ProviderType.aws, awsProvidersQueryString);
    }
    if (
      (!ocpProviders ||
        (onboardingStatus === FetchStatus.complete && !onboardingErrors)) &&
      ocpProvidersFetchStatus !== FetchStatus.inProgress
    ) {
      fetchProviders(ProviderType.ocp, ocpProvidersQueryString);
    }
    if (location.pathname !== prevProps.location.pathname) {
      window.scrollTo(0, 0);
    }
  }

  public componentWillUnmount() {
    this.appNav();
    this.buildNav();
  }

  public render() {
    return (
      <I18nProvider locale={this.state.locale}>
        <Routes />
        <ProvidersModal />
      </I18nProvider>
    );
  }
}

function buildNavigation() {
  const currentPath = window.location.pathname.split('/').slice(-1)[0];
  return [
    {
      title: 'Overview',
      id: '',
    },
    {
      title: 'Aws Details',
      id: 'aws',
    },
    {
      title: 'OpenShift Details',
      id: 'ocp',
    },
  ].map(item => ({
    ...item,
    active: item.id === currentPath,
  }));
}

const mapStateToProps = createMapStateToProps<AppOwnProps, AppStateProps>(
  (state, props) => {
    const awsProvidersQueryString = getProvidersQuery(awsProvidersQuery);
    const awsProviders = providersSelectors.selectProviders(
      state,
      ProviderType.aws,
      awsProvidersQueryString
    );
    const awsProvidersFetchStatus = providersSelectors.selectProvidersFetchStatus(
      state,
      ProviderType.aws,
      awsProvidersQueryString
    );

    const ocpProvidersQueryString = getProvidersQuery(ocpProvidersQuery);
    const ocpProviders = providersSelectors.selectProviders(
      state,
      ProviderType.ocp,
      ocpProvidersQueryString
    );
    const ocpProvidersFetchStatus = providersSelectors.selectProvidersFetchStatus(
      state,
      ProviderType.ocp,
      ocpProvidersQueryString
    );

    return {
      awsProviders,
      awsProvidersFetchStatus,
      awsProvidersQueryString,
      ocpProviders,
      ocpProvidersFetchStatus,
      ocpProvidersQueryString,
      onboardingErrors: onboardingSelectors.selectApiErrors(state),
      onboardingStatus: onboardingSelectors.selectApiStatus(state),
    };
  }
);

const mapDispatchToProps: AppDispatchProps = {
  history,
  fetchProviders: providersActions.fetchProviders,
};

export default hot(module)(
  compose(
    withRouter,
    connect(
      mapStateToProps,
      mapDispatchToProps
    )
  )(App)
);
