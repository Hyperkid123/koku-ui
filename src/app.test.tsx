import { shallow } from 'enzyme';
import React from 'react';

import { App, AppOwnProps } from './app';
import { FetchStatus } from './store/common';

const props: AppOwnProps = {
  fetchProviders: jest.fn(),
  awsProviders: null,
  awsProvidersFetchStatus: FetchStatus.none,
  awsProvidersQueryString: '',
  history: {
    push: jest.fn(),
    listen: jest.fn(),
  },
  location: null,
  ocpProviders: null,
  ocpProvidersFetchStatus: FetchStatus.none,
  ocpProvidersQueryString: '',
} as any;

test('does not fetch current user if not logged in', () => {
  shallow(<App {...props} />);
});

test('renders login if isLoggedIn is false', () => {
  const view = shallow(<App {...props} />);
  expect(view).toMatchSnapshot();
});

test('renders page layout', () => {
  const view = shallow(<App {...props} />);
  expect(view).toMatchSnapshot();
});
