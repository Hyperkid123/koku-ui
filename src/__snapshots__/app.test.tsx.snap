// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders login if isLoggedIn is false 1`] = `
<I18nProvider
  locale="en"
>
  <withRouter(Maintenance) />
</I18nProvider>
`;

exports[`renders page layout 1`] = `
<I18nProvider
  locale="en"
>
  <withRouter(Maintenance) />
</I18nProvider>
`;
